### DISCUSSION TITLE
Preserving trailing closure in Swift

### DISCUSSION THREAD

**Discussion by sberrevoets** [ast-grep/ast-grep#572] [Q&A] [answered]
In the following Swift snippet:

```swift
foo(a: a, b: b, c: c) { state in
	state.a = a
    state.b = b
}
```

`sg -p 'foo(a: $A, b: $B, c: $C) { $E in $F }' -l swift` yields

```
1│foo(a: a, b: b, c: c) { state in
2│    state.a = a
3│    state.b = b
4│}
```

Now I want to swap parameter `a` and `b`. My initial thought was ignoring the trailing closure would be fine: `sg -p 'foo(a: $A, b: $B, c: $C)' -l swift --rewrite 'foo(b: $B, a: $A, c: $C)'`, but this results in

```
1  │-foo(a: a, b: b, c: c) { state in
2  │-    state.a = a
3  │-    state.b = b
4  │-}
  1│+foo(c: c, a: a, b: b)
```

Then I tried `sg -p 'foo(a: $A, b: $B, c: $C) { $E in $F }' -l swift --rewrite 'foo(b: $B, a: $A, c: $C) { $E in $F }'` but that only preserved the first line of the closure:

```
1  │-foo(a: a, b: b, c: c) { state in
2  │-    state.a = a
3  │-    state.b = b
4  │-}
  1│+foo(b: c, a: a, c: c) { state in state.a = a }
```

I also tried `sg -p 'foo(a: $A, b: $B, c: $C) { $$$ }' -l swift --rewrite 'foo(c: $C, a: $A, b: $B) { $$$ }` but that doesn't match anything at all. Am I doing it wrong?

**Comment by HerringtonDarkholme** [comment, marked as answer]
Thanks for using ast-grep!

I think the solution here is 

```bash
sg -p 'foo(a: $A, b: $B, c: $C) { $E in $$$F }' -l swift --rewrite 'foo(b: $B, a: $A, c: $C) { $E in $$$F }'
```

However, looks like `$$$F` is not a valid identifier in Swift so ast-grep will not work here. It should be fixed by the CLI tool.


I will convert this issue to a bug report! Thank you!

**Reply by sberrevoets** [reply]
Awesome, thanks for the prompt reply! I think one problem here is that unless the CLI wants to preserve the existing spacing, a closure like

```swift
{ a in
	a.doSomething()
	a.doSomethingElse()
}
```

can't be one-lined without adding more syntax (a `;`). Just mentioning that as a heads up that there may be some complication involved.

Would you say the first version where it removes the closure altogether works as expected?

**Reply by HerringtonDarkholme** [reply]
It is even hard for me to tell if the first version is expected. 
The lack of Swift playground makes this hard... I will try adding one in the near future.

My guess here is that the first version skips matching lambda literal at all and replace the whole call expression.
See the bumped s-exp here.

```
(
  call_expression 
  (simple_identifier) 
  (
    call_suffix 
    (value_arguments) 
    (lambda_literal 
      type: (
        lambda_function_type (
          lambda_function_type_parameters (
            lambda_parameter name: (simple_identifier)))) 
      (statements (simple_identifier)))))

```

**Reply by HerringtonDarkholme** [reply]
For multiple line input on CLI, it is already supported. For example, in this case

```
sg -p 'foo(a: $A, b: $B, c: $C) { $E in
$F
$D
}'
```

Can match two statements with `F` and `D`

**Reply by sberrevoets** [reply]
Ah yes I didn't think about that, makes sense 💯 Sounds like your original comment is what I'd need since the number of lines in the closure is variable meaning I can't use your last suggestion.

Thanks for explaining and a great tool!

**Reply by HerringtonDarkholme** [reply]
Thanks for the kind words! I will work on swift playground this weekend. Stay tuned!

**Reply by HerringtonDarkholme** [reply]
Shipped in 0.10.1 🎉

**Reply by sberrevoets** [reply]
Awesome, that will be very helpful in the future! I'm not too familiar with tree sitter or the internals of this tool, does the playground help triage the issue?

**Reply by HerringtonDarkholme** [reply]
Playground is the main tool for debugging tree sitter.
I have also shipped Swift support in the web playground so it will be easier to triage issues 😁
